main.go

```diff
@@ -3,126 +3,158 @@ package main
+	flag.IntVar(&pingDuration, "ping-duration", 5, "Ping duration in seconds, 0 will disable pings to the tunnel")
 	flag.Parse()
 
+	customFormatter := new(logrus.TextFormatter)
+	customFormatter.FullTimestamp = true
+	logrus.SetFormatter(customFormatter)
+	logrus.SetLevel(logrus.DebugLevel)
+
 	if cloudURI == "" {
-		fmt.Printf("cloud-uri must be provided\n")
+		logrus.Errorf("cloud-uri must be provided\n")
 
 		os.Exit(1)
 	}
 
 	if apiKey == "" {
-		fmt.Printf("api-key must be provided\n")
+		logrus.Errorf("api-key must be provided\n")
 
 		os.Exit(1)
 	}
 
 	// setup tcp server
 	listener, err := net.Listen("tcp", listenURI)
 	if err != nil {
-		fmt.Printf("Error listening: %s\n", err.Error())
+		logrus.Errorf("Error listening: %s\n", err.Error())
 
 		os.Exit(1)
 	}
 	defer listener.Close()
 
-	fmt.Printf("Listening on: %s\n", listenURI)
+	if pingDuration > 0 {
+		logrus.Infof("Send pings enabled with an interval %v seconds\n", pingDuration)
+	} else {
+		logrus.Infof("Send pings disabled\n")
+	}
+
+	logrus.Infof("Listening on: %s\n", listenURI)
 
 	ctx, cancel := context.WithCancel(context.Background())
 
-	go func() {
-		defer cancel()
+	defer cancel()
 
-		for {
-			tcpConn, err := listener.Accept()
-			if err != nil {
-				fmt.Printf("Error accepting: %s\n", err.Error())
+	for {
+		tcpConn, err := listener.Accept()
+		if err != nil {
+			logrus.Errorf("Error accepting: %s\n", err.Error())
 
-				return
-			}
+			return
+		}
 
-			fmt.Printf("Received message from %s to %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr())
+		logrus.Debugf("(Connection from %s to %s) Starting\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr())
 
-			wsConn, _, err := websocket.DefaultDialer.Dial(cloudURI, http.Header{
-				"Authorization": []string{
-					"Bearer " + apiKey,
-				},
-				"X-Application-ID": []string{applicationID},
-			})
+		wsConn, _, err := websocket.DefaultDialer.Dial(cloudURI, http.Header{
+			"Authorization": []string{
+				"Bearer " + apiKey,
+			},
+			"X-Application-ID": []string{applicationID},
+		})
 
-			if err != nil {
-				fmt.Printf("Error dial into edge tunneling service: %s\n, cloudURI - %s\n", err.Error(), cloudURI)
+		if err != nil {
+			logrus.Errorf("(Connection from %s to %s) Error dial into edge tunneling service: %s\n, cloudURI - %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr(), err.Error(), cloudURI)
 
-				return
-			}
+			return
+		}
 
-			ctx, cancel := context.WithCancel(ctx)
+		_, cancel := context.WithCancel(ctx)
 
-			go func() {
-				for {
+		go func() {
+			chanBytes := make(chan []byte, 4096)
+			go func () {
+				for{
 					bytes := make([]byte, 4096)
 					length, err := tcpConn.Read(bytes)
 					if err != nil {
-						fmt.Printf("failed to read bytes from tcp connection: %s\n", err.Error())
+						logrus.Errorf("(Connection from %s to %s) failed to read bytes from tcp connection: %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr(), err.Error())
+						cancel()
```

Comment:
`cancel()` should be called as a defer function so we could make sure it could be called everytime.

Reply:
This is in different go routine, cancelling a connection based ctx.

Reply:
Not a big deal here. I will let you to decide if you want to move it or not.

Reply:
Done.